eosmetrics/emtr-aggregate-timer.c

```diff
@@ -0,0 +1,167 @@
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = emtr_aggregate_timer_finalize;
+}
+
+static void
+emtr_aggregate_timer_init (EmtrAggregateTimer *self)
+{
+}
+
+static void
+on_aggregate_timer_proxy_created_cb (GObject      *source_object,
+                                     GAsyncResult *result,
+                                     gpointer      user_data)
+{
+  EmtrAggregateTimer *self = EMTR_AGGREGATE_TIMER (user_data);
+  g_autoptr(GError) error = NULL;
+
+  self->timer_proxy =
+    emer_aggregate_timer_proxy_new_for_bus_finish (result, &error);
+
+  if (error)
+    g_warning ("Error creating aggregate timer: %s", error->message);
+
+  if (self->stopped && self->timer_proxy)
+    emer_aggregate_timer_call_stop_timer (self->timer_proxy, NULL, NULL, NULL);
+
+  g_object_unref (self);
+}
+
+static void
+on_server_aggregate_timer_started_cb (GObject      *source_object,
+                                      GAsyncResult *result,
+                                      gpointer      user_data)
+{
+  EmtrAggregateTimer *self = EMTR_AGGREGATE_TIMER (user_data);
+  EmerEventRecorderServer *event_recorder_server;
+  g_autofree gchar *timer_object_path = NULL;
+  g_autoptr(GError) error = NULL;
+  GDBusConnection *connection;
+  GDBusProxy *proxy;
+
+  event_recorder_server = EMER_EVENT_RECORDER_SERVER (source_object);
+  emer_event_recorder_server_call_start_aggregate_timer_finish (event_recorder_server,
+                                                                &timer_object_path,
+                                                                result,
+                                                                &error);
+
+  if (error)
+    {
+      g_warning ("Error creating aggregate timer: %s", error->message);
+      g_object_unref (self);
+      return;
+    }
+
+  proxy = G_DBUS_PROXY (source_object);
+  connection = g_dbus_proxy_get_connection (proxy);
+  emer_aggregate_timer_proxy_new (connection,
+                                  G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES |
+                                  G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS |
+                                  G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START,
+                                  g_dbus_proxy_get_name (proxy),
+                                  timer_object_path,
+                                  NULL,
+                                  on_aggregate_timer_proxy_created_cb,
+                                  g_object_ref (self));
+  g_object_unref (self);
+}
+
+EmtrAggregateTimer *
+emtr_aggregate_timer_new (EmerEventRecorderServer *dbus_proxy,
+                          GVariant                *event_id,
+                          GVariant                *aggregate_key,
+                          GVariant                *auxiliary_payload)
+{
+  EmtrAggregateTimer *self;
+
+  self = g_object_new (EMTR_TYPE_AGGREGATE_TIMER, NULL);
+
+  emer_event_recorder_server_call_start_aggregate_timer (dbus_proxy,
```

Comment:
A purist would say that this function should be called from `GObjectClass.constructed` and by extension all these variables should be GObject properties. I can see why one might not want to do that though... :-)

Reply:
This case I think it's ok because the app don't construct these directly, gets them from a function on another class